src/tdigest.rs

```diff
@@ -0,0 +1,269 @@
+    delta: f32,
+    buffer: [f32; 32], // used exclusively for add_value
+    index: u8, // see above comment
+    temp_centroids: Vec<Centroid>, // used exclusively for cluster_centroids to save on vector allocations
+}
+
+impl TDigest {
+    /// Creates a new empty T-Digest given a delta value.
+    pub fn new_empty(delta: f32) -> Self {
+        Self {
+            centroids: Vec::new(),
+            delta,
+            buffer: [0.0; 32],
+            index: 0,
+            temp_centroids: Vec::new(),
+        }
+    }
+
+    // the following is not necessary for the python package (you can only have one new function)
+    /// Creates a new T-Digest from an input vector of unsorted values and a delta value. Equivalent to creating an empty T-Digest and merging in a vector of values.
+    // pub fn new_from_vec(v: Vec<f32>, delta: f32) -> Self {
+    //     let mut digest = Self {
+    //         centroids: Vec::new(),
+    //         delta,
+    //         buffer: [0.0; 32],
+    //         index: 0,
+    //         temp_centroids: Vec::new(),
+    //     };
+    //     digest.merge_vec_unsorted(v);
+    //     digest
+    // }
+
+    // apparently, this is not a std function
+    fn clamp(v: f32, lo: f32, hi: f32) -> f32 {
+        if v > hi {
+            hi
+        } else if v < lo {
+            lo
+        } else {
+            v
+        }
+    }
+
+    // does what it says. probably not necessary since it's only used once
+    fn vals_to_centroids(v: Vec<f32>) -> Vec<Centroid> {
+        v.iter().map(|x| Centroid::new(*x, 1)).collect()
+    }
+
+    // THIS FUNCTION WILL DIE IF THE INTERNAL CENTROIDS AREN'T SORTED BY MEAN!!!!!
+    // takes an unoptimized list of centroids and condenses it down to fit within the digest's delta value
+    fn cluster_centroids(&mut self) {
+        if self.centroids.len() == 0 {
+            return
+        }
+        self.temp_centroids.clear(); // used as a buffer to store the new centroids while accessing the old
+        let weight_sum: i32 = self.centroids.iter().map(|x| x.weight).sum(); // sum of weights of all centroids (equivalent to total number of values added to digest)
+        let mut k_limit: f32 = 1.0; // if you read the gresearch article, it's actually slightly wrong. the new group of centroids doesn't start at the one that pushes the delta k past 1, it starts at each k whole number (so that you don't end up with >delta centroids)
+        let mut q_lower: f32 = 0.0; // lower quantile limit. feel like there's a small optimization to be had here but i'm not seeing it.
+        let mut q_limit = inverse_scale(k_limit, self.delta); // precomputed as a max quantile value here & after each new centroid is completed so that you don't have to compute new k values with every single iteration
+        let mut current_centroid = self.centroids[0]; // also used as a buffer, stores the aggregated centroid that is to be added to temp_centroids
+        for i in 1..self.centroids.len() {
```

Comment:
that would be cleaner, but i feel like this is not super unreadable and it saves on One Whole Iteration. unless vector accesses are really slow and i just didn't know that, in which case let me know.

Reply:
Another way to write this and avoid the initial slicing is using the `fold` iterator. For example something like this, it doesn't use `self.temp_centroids`, and does allocate a new vec each time the function is called, but maybe something you can build upon:
```rust
        self.centroids = self.centroids.iter().fold(Vec::new(), |mut acc, centroid| {
            // In the first iteration, just add the centroid to the acc
            let Some(prev_centroid) = acc.last_mut() else {
                acc.push(*centroid);
                return acc;
            };
            // Then either merge the current and previous centroids, or update the limits
            // and push the centroid as it is.
            let q_upper =
                q_lower + (prev_centroid.weight + centroid.weight) as f32 / weight_sum as f32;
            if q_upper <= q_limit {
                prev_centroid.merge(centroid);
            } else {
                q_lower += prev_centroid.weight as f32 / weight_sum as f32;
                k_limit += 1.0;
                q_limit = inverse_scale(k_limit, self.delta);
                acc.push(*centroid);
            }
            acc
        });

```

Reply:
while it does look nicer, even if i found a way to avoid the vector allocation i can't see it providing a worthwhile performance benefit. plus (at least to me) a for loop is a bit easier to wrap my head around. since the initial slice isn't actually manipulating any vector elements, i can't imagine it taking up any significant portion of the function's runtime